### ISSUE TITLE
Tern js doesn't auto complete when expected

### ISSUE THREAD

**Issue by StefanICT** [marijnh/tern_for_vim#89]
I added two images of my problem. The first shows that Vim doesn't show to docs for whatever function, but when I prepend "this." it does? Why does this happen, am I missing some .tern-project setting? 

![screen shot 2015-03-31 at 23 41 07](https://cloud.githubusercontent.com/assets/3540810/6929969/7ad3317c-d7ff-11e4-9034-61396f88ed21.png)
![screen shot 2015-03-31 at 23 41 21](https://cloud.githubusercontent.com/assets/3540810/6929972/7fdb236e-d7ff-11e4-9d77-c6a87161e30b.png)

My .tern-project
````
{"libs": ["browser", "ecma5"] }
````

**Comment by marijnh** [comment]
This sounds a bit like #516, but once again, I can not reproduce it -- I do see `(fn)` after the name when I complete a global function (I had to add a second function starting with `what`, because my vim immediately completes when there is only a single match).

**Comment by StefanICT** [comment]
Could you show me a picture or something of the result you get? Also do you have extra stuff in the vimrc that could be the cause that mine completion is not working? I just reinstalled tern and YoucompleteMe but no luck. I love this plugin and I need it ;) Or show me how I can debug it?

**Comment by marijnh** [comment]
My `.vimrc` is simply this:

```
execute pathogen#infect()
syntax on
filetype plugin indent on
```

I just moved to a new machine and installed vim specifically to try to reproduce this issue. So the only plugin that I have installed is tern_for_vim. Does disabling other plugins make a difference?

**Comment by StefanICT** [comment]
I only got YouCompleteMe and without it does't work either. I also moved to new machine. My vimrc isn't much more then that:

````
set nocompatible

execute pathogen#infect()
syntax on
filetype plugin indent on

"Habit breaking, habit making
nnoremap <Up> :echoe "use k" <CR>
nnoremap <Down> :echoe "use j" <CR>
nnoremap <Left> :echoe "use h" <CR>
nnoremap <Right> :echoe "use l" <CR>

set rnu 
set nu
set ai
"Backspace not working for some reason this fixes it
set backspace=indent,eol,start
colo candycode

````

**Comment by marijnh** [comment]
Strange. Which Tern version are you using? (See the top of `node_modules/tern/package.json`)

**Comment by StefanICT** [comment]
````
"name": "tern",
  "license": "MIT",
  "version": "0.10.0",
  "author": {
    "name": "Marijn Haverbeke",
    "email": "marijnh@gmail.com"
  },
````

**Comment by marijnh** [comment]
Here's what completion on a simple file looks like, using Tern 0.10.0 and vim 7.4:

![image](https://cloud.githubusercontent.com/assets/144427/7033041/75300c40-dd77-11e4-956c-3bc9afa8b96f.png)

**Comment by StefanICT** [comment]
So the setup of the situation is:

main.js
    -code
.tern-project
  - {} // just empty

Right?

**Comment by marijnh** [comment]
I didn't create a `.tern-project`, just a `foo.js` in my home dir, but the effect should be the same.

You could see what you get when you start your tern server manually from the command line with the `--verbose` flag (kill existing servers first). Do the completions really have `"?"` as the type? If not, something is going wrong on the vim side. Next step would be to try and dig down into the the `var type = val.getType()` line in `findCompletions` in `tern/lib/tern.js`, and figure out why it isn't returning a type -- but this library is unfortunately not very easy to debug without a lot of background knowledge.

**Comment by StefanICT** [comment]
I appreciate the help. So i am digging right now. I will keep you posted, could be tomorrow though

**Comment by barwinco** [comment]
@StefanICT

> I only got YouCompleteMe and without it does't work either.

This means that you've tried to complete 'what' manually with ```C-x C-o``` and it does not work? Strange, because ```C-x C-o``` works for me.

YCM and neocomplete do not complete global names without dot, 
I have created a dictionary for neocomplete with ```document```, ```window```, ```prompt```, ```function``` and stuff like that, which works fine for me.

**Comment by StefanICT** [comment]
@barwinco w00t w000t...!!!!

I just tried disabling YCM and in the docs I read the <C-space> activates the completion and then it works! I guess vim is just in the way or something? I guess I need to find a way to let YCM handle the completion or something!

On the right track!!!

**Comment by StefanICT** [comment]
I think I found the solution!

YouCompleteMe gives two types of completion. I want the semantics and stopped the other. YCM got this option to set triggers:

````
let g:ycm_semantic_triggers =  {
  \   'c' : ['->', '.'],
  \   'objc' : ['->', '.'],
  \   'ocaml' : ['.', '#'],
  \   'cpp,objcpp' : ['->', '.', '::'],
  \   'perl' : ['->'],
  \   'php' : ['->', '::'],
  \   'cs,java,javascript,d,python,perl6,scala,vb,elixir,go' : ['.'],
  \   'vim' : ['re![_a-zA-Z]+[_\w]*\.'],
  \   'ruby' : ['.', '::'],
  \   'lua' : ['.', ':'],
  \   'erlang' : [':'],
  \ }
````

I have to add a python regex to match any 3 letters, then it is fixed and this can be closed now I guess.

**Comment by ruanyl** [comment]
@StefanICT Hi, what regex did you add to make javascript auto trigger to work? I have the similar issue as you.

**Comment by talkw-liu** [comment]
Why it doesn't show the write function for me? 
![1533888332 1](https://user-images.githubusercontent.com/11623379/43946454-e2e9a036-9cb7-11e8-9ee2-9f8714f96df8.jpg)

My .tern-project:
{
  "libs": [
      "browser",
      "underscore",
      "jquery"
    ],  

  "plugins": {
        "node": {}, 
        "requirejs": {
            "baseURL": "./",
            "paths": {}
        }   
    }   
}

**Comment by marijnh** [comment]
@talkw-liu That's a different issue, fixed by d43faf3
